### PR TITLE
feat: [WD-16972] TLS user management spike.

### PR DESCRIPTION
**Pull request by Kxiru**
## Done
- Created "Create Identity" Button
- Create modal to create new TLS fine grained identities.
- A few new functions and a new LXDTrustToken type.

## User Management Spike
[✔️] A user needs to be able to create a TLS user (Creates a pending user)
[✔️] A user needs to be able to destroy a TLS user ([WD-16894](https://github.com/canonical/lxd-ui/pull/1008))

Fixes:
- Inability to create TLS fine grained tokens through the UI.

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - [List the steps to QA the new feature(s) or prove that a bug has been resolved]

## Screenshots
image
Screenshot from 2024-12-10 22-47-10

[WD-16894]: https://warthogs.atlassian.net/browse/WD-16894?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/api/auth-identities.tsx [hunk 1]
```diff
@@ -74,3 +74,20 @@ export const deleteOIDCIdentities = (
       .catch(reject);
   });
 };
+
+export const fetchIdentityDetails = (
```
Comment (mas-who) [inline src/api/auth-identities.tsx:78]: I think `createFineGrainedTlsIdentity` might be a more appropriate name for this function, wdyt?

File: src/types/permissions.d.ts [hunk 2]
```diff
@@ -30,3 +30,12 @@ export interface IdpGroup {
   name: string;
   groups: string[]; // these should be names of lxd groups
 }
+
+export interface LxdTrustToken {
```
Comment (mas-who) [inline src/types/permissions.d.ts:34]: this type should be named `LxdFineGrainedTlsIdentity`, the token is what we will generate in the browser.

File: src/types/permissions.d.ts [hunk 3]
```diff
@@ -30,3 +30,12 @@ export interface IdpGroup {
...
+  client_name: string;
+  addresses: string[];
+  expires_at: string;
+  fingerprint: string;
+  type: string;
```
Comment (mas-who) [inline src/types/permissions.d.ts:39]: we can be more specific with the `type` field since it has limited number of possible values

File: src/util/helpers.tsx [hunk 4]
```diff
@@ -335,3 +336,19 @@ export const getDefaultStoragePool = (profile: LxdProfile) => {
...
+    secret: tokenPreResponse.secret,
+    expires_at: tokenPreResponse.expires_at,
+    type: tokenPreResponse.type,
+  };
+
+  const jsonString = JSON.stringify(tokenDetails);
+  const base64Encoded = btoa(jsonString);
+
+  return base64Encoded;
+};
```
Comment (mas-who) [inline src/util/helpers.tsx:354]:
```suggestion
export const encodeBase64Token = (data: unknown) => {
  const jsonString = JSON.stringify(data);
  return btoa(jsonString);
};
```

We shouldn't couple a general concept like base64 encoding to the tls identities data.

File: src/sass/_instance_detail_page.scss [hunk 5]
```diff
@@ -10,7 +10,8 @@
 
 .create-instance-from-snapshot-modal,
 .duplicate-instances-modal,
-.create-image-from-instance-modal {
+.create-image-from-instance-modal,
+.create-tls-identity {
```
Comment (mas-who) [inline src/sass/_instance_detail_page.scss:14]: this feels a bit off adding identity related styling here, since the file is supposed to be dedicated for the instance detail page. Looking at the styles, it's a general modal dialog style override, I'd suggest move the styling into `styles.scss` and potentially come up with a more general class name. wdyt?

File: src/pages/permissions/PermissionIdentities.tsx [hunk 6]
```diff
@@ -50,6 +53,12 @@ const PermissionIdentities: FC = () => {
   const [searchParams] = useSearchParams();
   const [selectedIdentityIds, setSelectedIdentityIds] = useState<string[]>([]);
   const { hasAccessManagementTLS } = useSupportedFeatures();
+  const isSmallScreen = useSmallScreen();
+  const { openPortal, closePortal, isOpen, Portal } = usePortal();
+
+  const handleClose = () => {
```
Comment (mas-who) [inline src/pages/permissions/PermissionIdentities.tsx:59]: no need to have this function here, we can just pass `closePortal` to `onClose` below

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      fetchIdentityDetails(values.clientName)
+        .then((tokenPreResponse) => {
+          const tempToken = base64EncodeTokenDetails(tokenPreResponse);
+
+          setIsLoadingToken(false);
+          setToken(tempToken);
+        })
+        .catch((err) => {
+          setIsLoadingToken(false);
+          console.error("Error fetching token:", err);
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:37]: please remove the console log and use toast notification for error and success messages

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+  return (
+    <Modal
+      close={onClose}
+      className="create-tls-identity"
+      title="Generate Trust Token"
+      buttonRow={
+        <>
+          {token ? (
+            <ActionButton
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:50]: We don't need a refresh button to update identities list, instead we should use react query invalidation

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            </ActionButton>
+          ) : (
+            <ActionButton
+              appearance="positive"
+              className="u-no-margin--bottom"
+              onClick={() => void formik.submitForm()}
+              disabled={
+                formik.values.clientName.length === 0 ||
+                isLoadingToken ||
+                token != null
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:65]: Instead of two buttons, we can have one and adjust the onclick logic as well as the button label based on if a token has been generated

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+              }
+            >
+              Generate Token
+            </ActionButton>
+          )}
+        </>
+      }
+    >
+      {isLoadingToken && <Loader text="Loading token..." />}
+      {!isLoadingToken && !token && (
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:75]: We should disble the input instead of hiding it when making the API call, this could reduce layout shift in the modal. The loading state can be indicated by the action button instead of the dedicated loader

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        <Input
+          id="clientName"
+          type="text"
+          label="Client Name"
+          onBlur={formik.handleBlur}
+          onChange={formik.handleChange}
+          value={formik.values.clientName}
+        />
+      )}
+      {!isLoadingToken && token && (
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:85]: as long as the token exist, we won't be in the loading state

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        <div>
+          <p>
+            Please copy and paste your token for future reference as it will not
+            be available again.
+          </p>
+          <code>{token}</code>
+        </div>
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:92]: I think the div is not necessary here, we can just use a fragment

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+import Loader from "components/Loader";
+import { fetchIdentityDetails } from "api/auth-identities";
+import { base64EncodeTokenDetails } from "util/helpers";
+
+interface Props {
+  onClose: () => void;
+}
+
+const CreateIdentityModal: FC<Props> = ({ onClose }) => {
+  const [isLoadingToken, setIsLoadingToken] = useState(false);
```
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:13]: maybe this state can simply be `isLoading` as it is the only loading state for the modal

**Review by mas-who — COMMENTED** [review]
Made some comments, I agree with @edlerd that this PR can actually be merged after some improvements. Although I'm also happy to not merge it and keep the PR as a reference for future, depending on if you have time to take it to completion

File: src/sass/_instance_detail_page.scss (hunk 5 above)
Comment (Kxiru) [reply src/sass/_instance_detail_page.scss:14]: Oh right, I see what you mean.
Yes, it is a general style. I have moved it into styles.scss for now, however do you think forms.scss would be appropriate? Not too sure, because although modals include forms, it is not directly relevant to forms themselves. Perhaps styles.scss is best as suggested.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 7 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:37]: Will we need a success message if the token is being presented right in front of us and we are already in a modal? I note this because if one exits the modal after generating a token by pressing refresh, we wouldn't see the toast notification anyway. Also a token would only be generated if the operation was successful.

Comment (mas-who) [reply src/pages/permissions/CreateIdentityModal.tsx:37]: that's a interesting one, I think perhaps have some copy in the modal to indicate that an TLS identity was created. I'm not entirely confident that is the correct design though, probably a good topic to discuss during standup. The toast notification would be somewhat hidden when the modal is open anyway...

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 9 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:65]: I think this gets addressed by invalidating the queries :)

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 10 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx]: Hmm. Won't this approach also result in layout shift because at one point the tokens are there and at the next, they aren't?
In my recent commit, I took out the loader state entirely, but I have still left the input box and token box there and moved the loading feedback to the action button. Let me know what you think?

Comment (mas-who) [reply src/pages/permissions/CreateIdentityModal.tsx]: the layout shift would be during the loading state for the modal. If we don't show the loader and just disable the input, then during the loading state, the modal will not have its size adjusted

Comment (mas-who) [reply src/pages/permissions/CreateIdentityModal.tsx]: so what you've done now seems good :+1:

File: src/types/permissions.d.ts [hunk 14]
```diff
@@ -30,3 +30,12 @@ export interface IdpGroup {
   name: string;
   groups: string[]; // these should be names of lxd groups
 }
+
+export interface LxdFineGrainedTlsIdentity {
+  client_name: string;
+  addresses: string[];
+  expires_at: string;
+  fingerprint: string;
+  type: "oidc" | "Client certificate (pending)" | "Client certificate";
```
Comment (mas-who) [inline src/types/permissions.d.ts:39]: I think `type` can't be `oidc` in this case since this data gets returned from the `.../tls` endpoint. On second thought, maybe we can change the type name to `LxdTlsIdentity` instead, it's less verbose

File: src/pages/permissions/CreateIdentityModal.tsx
Comment (mas-who) [inline src/pages/permissions/CreateIdentityModal.tsx:1]: Code in this file looks pretty good now, thanks for the changes. I believe things will still change here based on our discussion earlier

Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:1]: Sounds good, thank you!

File: src/types/permissions.d.ts (hunk 14 above)
Comment (Kxiru) [reply src/types/permissions.d.ts:39]: Surely the type name should remain LxdFineGrainedTlsIdentity as "LxdTlsIdentity is too broad. It does not consider Unrestricted TLS identities.

Comment (mas-who) [reply src/types/permissions.d.ts]: I've done some digging look at the API [docs](https://documentation.ubuntu.com/lxd/en/latest/api/#/), please refer to it for the following observation:

1. The naming for the type is wrong, since `LxdFineGrainedTlsIdentity` is a very specific response data struct only returned by making a request to `POST /1.0/auth/identities/tls`. The response does not represent an "official" identity data structure in LXD, and it includes details like `addresses`, `expires_at`, `fingerprint` and `secret` etc because those are required to authenticate a TLS identity later. Instead, we should think of a specific type name for the response data from this endpoint, perhaps what you had earlier was a closer fit e.g. `TlsIdentityTokenDetail`.

2. The official identity data structure from the LXD backend can be retrieved from `GET /1.0/auth/identities?recursion=1`, which is already represented by the existing `LxdIdentity` type. Can you please augment the existing type by looking at the response data from the documentation, add what is missing and also make the `type` field more strict (you will need to look at the network request for this since the documentation isn't all comprehensive).

3. After looking at the documentation for the `POST /1.0/auth/identities/tls` endpoint, I see we can set a `certificate` field in the request payload, instead of `token`. Please would you investigate what happens if we set `certificate` to `true`. I suspect we can potentially skip the create certificates in the UI step with this (for future reference).

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+          )}
+          <ActionButton
+            appearance="positive"
+            className="u-no-margin--bottom"
+            onClick={token ? onClose : () => void formik.submitForm()}
+            disabled={formik.values.clientName.length === 0 || isLoading}
+            loading={isLoading}
+          >
+            {token ? `Close` : `Generate Token`}
+          </ActionButton>
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:80]: I think the close should not have a positive appearance, and it doesn't need the disabling or loading props.

So there is no overlap in the props, it seems better to have two distinct buttons.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+          {token && (
+            <Button
+              className="u-no-margin--bottom"
+              onClick={() => {
+                navigator.clipboard.writeText(token);
+              }}
+              type="button"
+              hasIcon
+            >
+              <Icon name="copy" />
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:69]: This should have a copy label as well, to be discoverable to screen readers.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+          type="text"
+          label="Client Name"
+          onBlur={formik.handleBlur}
+          onChange={formik.handleChange}
+          value={formik.values.clientName}
+        />
+      )}
+      {token && (
+        <>
+          <p>Identity has been created.</p>
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:102]: Can this be the heading of the modal? Idea is replacing the "Generate trust token" copy once we are at this step.

File: src/pages/permissions/PermissionIdentities.tsx [hunk 18]
```diff
@@ -264,6 +274,17 @@ const PermissionIdentities: FC = () => {
...
+              <Button
+                appearance="positive"
+                className="u-float-right u-no-margin--bottom"
+                onClick={openPortal}
+                hasIcon={!isSmallScreen}
+              >
+                {!isSmallScreen && <Icon name="plus" light />}
+                <span>Create TLS Identity</span>
+              </Button>
+            </PageHeader.BaseActions>
```
Comment (edlerd) [inline src/pages/permissions/PermissionIdentities.tsx:287]: All of this logic, with the button, mounting the modal and referencing the new CreateIdentityModal should move to a CreateTLSIdentityButton component, so we have contained the logic in a single place. This PermissionIdentities component responsibility should be the listing, not the action details.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+          <p>The token below can be used to log in as the new identity.</p>
+
+          <Notification
+            severity="caution"
+            title="Make sure to copy the token now as it will not be shown again."
+          ></Notification>
+
+          <div className="token-code-block">
+            <code>{token}</code>
+          </div>
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:112]: We should reuse the [CodeSnippet](https://canonical.github.io/react-components/?path=/story/components-codesnippet--wrap-lines) and avoid a custom styling for it. The custom styling creates more code and is missing margins.

File: src/util/helpers.tsx [hunk 20]
```diff
@@ -335,3 +336,8 @@ export const getDefaultStoragePool = (profile: LxdProfile) => {
     });
   return rootStorage ? rootStorage.pool : "";
 };
+
+export const base64EncodeTokenDetails = (data: LxdFineGrainedTlsIdentity) => {
```
Comment (edlerd) [inline src/util/helpers.tsx:340]: I think we should make this a bit more generic, i.e. call it `base64EncodeObject(data: Object)` so it can be reused in other contexts.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+  const [error, setError] = useState(false);
+  const toastNotify = useToastNotification();
+  const queryClient = useQueryClient();
+
+  const formik = useFormik({
+    initialValues: {
+      clientName: "",
+    },
+    onSubmit: (values) => {
+      setIsLoading(true);
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:33]: You can query `formik.isSubmitting`, no need for a custom state. The formik one is automatically toggled during submit.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+      createFineGrainedTlsIdentity(values.clientName)
+        .then((tokenPreResponse) => {
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:36]:
```suggestion
        .then((response) => {
```
What is a token pre response?

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+          const tempToken = base64EncodeTokenDetails(tokenPreResponse);
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:37]:
```suggestion
          const encodedToken = base64EncodeTokenDetails(tokenPreResponse);
```

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+          >
+            {token ? `Close` : `Generate Token`}
+          </ActionButton>
+        </>
+      }
+    >
+      {error && (
+        <Notification
+          severity="negative"
+          title="Fine grained identity token creation failed"
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:87]:
```suggestion
          title="Token creation failed"
```

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+      )}
+      {!token && (
+        <Input
+          id="clientName"
+          type="text"
+          label="Client Name"
+          onBlur={formik.handleBlur}
+          onChange={formik.handleChange}
+          value={formik.values.clientName}
+        />
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:98]: We are creating an identity, this should be called "Identity name", where does the "client name" idea come from?

**Review by edlerd — COMMENTED** [review]
This already works well. Some ideas below to streamline code and appearance.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 15 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:80]: Hmm...
Merging the close and submit funcitonality into one button was an earlier suggestion. Given you are not a fan of the "positive" appearance, do you recommend I use a normal button, or conditionally make the button appearance negative / null?

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 19 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:112]: CodeSnippet does not currently support strings without breaks. I have raised this in the Vanilla channel. There are some workarounds however these also require custom styling. Please see (this thread)[https://chat.canonical.com/canonical/pl/44y5kzsagfnsmfyja8hi5wwt6e] and let me know your thoughts.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 15 above)
Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:80]: I think the default appearance, same as the copy button, would be fine for the close button.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 19 above)
Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:112]: I see, it doesn't wrap long strings. Then, let's add the default margin below the element to fix the spacing.

Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:112]: image

Something like this, but not as hacky, better with a class and applying the styling via scss.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 25 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:98]: This is the terminology that the API and documentation uses. Yes, when the identity surfaces, it appears as the "Identity name", but I believe there is more to it than just the name and that staying true to the API may avoid potential issues in the future if the API changes against our mental model of "Identity name".

Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:107]: Right good points. I see why the docs are doing it. They are tailoring towards a new command line client. I think in the UI, the target audience is a bit different. A user might create the token for a new client, but it could also be for another person (think an admin creating the token for a colleague). In the API layer we use, this is just "name". I agree, a "Name" label would be confusing here when creating a token. The entity that is created is referred to as "Identity" in the UI. So maybe "Identity name" is the better description to cover all target use cases. Wdyt?

Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:107]: Sure, I think for the sake of the end user we can change client name to identity name. I do think that "Client name" allows for the user to think about their nming decision a little more, whilst "name" just implies a casual identity. To me, it's like the difference between "James's token" and "token 1" or "Chrome browser" and "laptop", because "client" implies the user should think more about how, why and where they are using it, as opposed to just what to call it. Especially as there is no renaming function and poor naming can make for confusing identities. Thoughts?

Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:107]: I think we should be consistent with the identity name as that is how the entity is presented after creation.

Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:107]: Alrighty! "Identity name" it is.

File: src/pages/permissions/PermissionIdentities.tsx [hunk 26]
```diff
@@ -181,7 +182,9 @@ const PermissionIdentities: FC = () => {
                 <Icon name="user-group" />
               </Button>
               {hasAccessManagementTLS && (
-                <DeleteIdentityBtn identity={identity} />
+                <>
+                  <DeleteIdentityBtn identity={identity} />
+                </>
```
Comment (edlerd) [inline src/pages/permissions/PermissionIdentities.tsx:187]: Why this change? Can it be reverted to keep it simple?

File: src/util/permissionIdentities.tsx [hunk 27]
```diff
@@ -5,6 +5,11 @@ export type ChangeSummary = Record<
   { added: Set<string>; removed: Set<string>; name: string }
 >;
 
+export type FineGrainedTLSFormValues = {
+  name: string;
+  publicity: boolean;
```
Comment (edlerd) [inline src/util/permissionIdentities.tsx:10]:
```suggestion
```
I think this is not a valid field and unused.

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+              title="Copy token"
+              className="u-no-margin--bottom"
+              onClick={() => {
+                void navigator.clipboard.writeText(token);
+              }}
+              type="button"
+              hasIcon
+            >
+              <Icon name="copy" />
+            </Button>
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:64]: Can we add some feedback after this is used?
A [good example interaction is found in AMS](https://github.com/canonical/ams/blob/main/ui/templates/js/components/ui/CopyToClipboard.tsx#L10):

File: src/pages/permissions/CreateIdentityModal.tsx [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+        .then((response) => {
+          const encodedToken = base64EncodeObject(response);
+          setToken(encodedToken);
+
+          void queryClient.invalidateQueries({
+            queryKey: [queryKeys.identities],
+          });
+        })
+        .catch(() => {
+          setError(true);
```
Comment (edlerd) [inline src/pages/permissions/CreateIdentityModal.tsx:40]: This never gets unset to false. Maybe can do it on submit?

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 16 above)
Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:63]: This is still not resolved.

Please do not resolve comments yourself. Better add comment with a reference to the commit that fixes the issue that was pointed out.

**Review by edlerd — CHANGES_REQUESTED** [review]
A couple of issues remaining below.

File: src/util/permissionIdentities.tsx (hunk 27 above)
Comment (Kxiru) [reply src/util/permissionIdentities.tsx:10]: I will remove the entire type as the rename functionality is not being introduced in this PR.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 29 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:40]: I have now added it on submit, but what about a handleClose function that sets it to false and calls onClose?

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 16 above)
Comment (Kxiru) [reply src/pages/permissions/CreateIdentityModal.tsx:63]: Hi David,
You asked for a label to be added for screen readers, and I added an aria-label. Was there anything else you wanted to add?
I resolve comments as I make code changes to keep track of my progress and minimise clutter on the screen which can be confusing.

**Comment by Kxiru** [comment]
>This is still not resolved.
>Please do not resolve comments yourself. Better add comment with a reference to the commit that fixes the issue that was pointed out.


Hi David,
You asked for a label to be added for screen readers, and I added an aria-label. Was there anything else you wanted to add?
I resolve comments as I make code changes to keep track of my progress and minimise clutter on the screen which can be confusing.

File: src/pages/permissions/CreateIdentityModal.tsx (hunk 29 above)
Comment (edlerd) [reply src/pages/permissions/CreateIdentityModal.tsx:42]: I suspect on close/reopen the modal gets rendered with initial state. It should not keep state from being opened previously.

**Review by edlerd — APPROVED** [review]
QA and code LGTM, good stuff, thx for adding this feature :)
